File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/AlertPresetPersistence.kt
package expo.modules.vescapecore.telemetry

import expo.modules.vescapecore.alerts.resolveConfigRelativeBase
import org.json.JSONObject
import java.util.UUID
import kotlin.math.floor

/** Native owns the preset selection and generated rules in one Room transaction.
 * @parity /modules/vescape-core/ios/telemetry/AlertPresetPersistence.swift
 * @parity /src/modules/alerts/lib/alertPresets.ts
 */
internal object AlertPresetPersistence {
  private val definition by lazy {
    JSONObject(checkNotNull(javaClass.getResourceAsStream("/alert-preset-definitions.json"))
      .bufferedReader().use { it.readText() })
  }
  val metrics = listOf("speed", "duty", "motor-temp", "controller-temp", "battery")
  private val levels = setOf("off", "safe", "normal", "minimal", "custom")
  private fun json(value: Any?) = JSONObject(value as? Map<*, *> ?: emptyMap<String, Any>())
  private fun roundTenth(value: Double) = floor(value * 10 + 0.5) / 10

  /** Older JS skipped unresolved matched specs. Restore only missing relationships, never replace rider state. */
  suspend fun repairMissingRelations(dao: TelemetryDao, boardId: String): Boolean {
    val settings = dao.getBoardSettings(boardId).associate { it.key to decodeSettingJson(it.valueJson) }
    val selection = json(settings["alertPreset"]); val match = json(settings["matchBoardConfig"])
    val existingIds = dao.getAlertRules(boardId).map { it.id }.toSet()
    var changed = false
    for (metric in metrics.filter { match.optBoolean(it) }) {
      for (rule in generate(dao, boardId, metric, selection.optString(metric, "off"), settings)) {
        if (rule.thresholdKind == "config-relative" && rule.id !in existingIds) { dao.upsertAlertRule(rule); changed = true }
      }
    }
    return changed
  }

  suspend fun regenerateChanged(dao: TelemetryDao, boardId: String, before: Map<String, Any?>) {
    val after = dao.getBoardSettings(boardId).associate { it.key to decodeSettingJson(it.valueJson) }
    val previous = json(before["alertPreset"]); val next = json(after["alertPreset"])
    val previousMatch = json(before["matchBoardConfig"]); val nextMatch = json(after["matchBoardConfig"])
    for (metric in metrics) {
      if (previous.optString(metric, "off") != next.optString(metric, "off") || previousMatch.optBoolean(metric) != nextMatch.optBoolean(metric) ||
        (metric == "speed" && ((before["topSpeedKmh"] as? Number)?.toDouble() != (after["topSpeedKmh"] as? Number)?.toDouble() || previous.optString("speedUnitSystem") != next.optString("speedUnitSystem"))) ||
        (metric == "battery" && before["batteryConfig"] != after["batteryConfig"])) regenerate(dao, boardId, metric)
    }
  }

  suspend fun regenerate(dao: TelemetryDao, boardId: String, onlyMetric: String? = null) {
    val settings = dao.getBoardSettings(boardId).associate { it.key to decodeSettingJson(it.valueJson) }
    val selection = json(settings["alertPreset"])
    for (metric in onlyMetric?.let { listOf(it) } ?: metrics) {
      dao.getAlertRules(boardId).filter { it.controlId == metric && it.source == "preset" }
        .forEach { dao.deleteAlertRule(boardId, it.id) }
      generate(dao, boardId, metric, selection.optString(metric, "off"), settings)
        .forEach { dao.upsertAlertRule(it) }
    }
  }

  suspend fun apply(dao: TelemetryDao, boardId: String, metric: String, action: String, level: String?, matchBoardConfig: Boolean?) {
    require(metric in metrics) { "Unknown alert preset metric" }
    require(dao.getBoard(boardId)?.deletedAt == null && dao.getBoard(boardId) != null) { "Board not found" }
    val settings = dao.getBoardSettings(boardId).associate { it.key to decodeSettingJson(it.valueJson) }.toMutableMap()
    val selection = json(settings["alertPreset"])
    metrics.forEach { if (selection.optString(it) !in levels) selection.put(it, "off") }
    val now = System.currentTimeMillis()
    var customized = emptyList<AlertRuleEntity>()
    when (action) {
      "select" -> { require(level in levels && level != "custom") { "Invalid preset level" }; selection.put(metric, level) }
      "customize" -> {
        repairMissingRelations(dao, boardId)
        if (selection.optString(metric) == "custom") return
        customized = dao.getAlertRules(boardId).filter { it.controlId == metric && it.source == "preset" }.mapNotNull { rule ->
          val base = if (rule.thresholdKind == "config-relative") resolveBase(dao, boardId, rule.configFieldId) ?: return@mapNotNull null else null
          rule.copy(id = UUID.randomUUID().toString(), source = null,
            threshold = if (base != null) base + (rule.thresholdOffset ?: return@mapNotNull null) else rule.threshold,
            thresholdMax = if (base != null) rule.thresholdMaxOffset?.let { base + it } else rule.thresholdMax,
            thresholdKind = "fixed", configFieldId = null, thresholdOffset = null, thresholdMaxOffset = null)
        }
        selection.put(metric, "custom")
      }
      "discard-custom" -> {
        dao.getAlertRules(boardId).filter { it.controlId == metric && it.source != "preset" }.forEach { dao.deleteAlertRule(boardId, it.id) }
        selection.put(metric, "normal")
      }
      "match-board-config" -> {
        require(definition.getJSONObject("match").has(metric) && matchBoardConfig != null) { "Metric cannot match board config" }
        val match = json(settings["matchBoardConfig"]).put(metric, matchBoardConfig)
        dao.upsertBoardSetting(BoardSettingEntity(boardId, "matchBoardConfig", match.toString(), now))
      }
      else -> error("Unknown alert preset action")
    }
    if (metric == "speed" && action in setOf("select", "discard-custom") && selection.optString(metric) !in setOf("off", "custom")) {
      val units = dao.getAppSetting("unitSystem")?.let { decodeSettingJson(it.valueJson) }
      selection.put("speedUnitSystem", if (units == "imperial") "imperial" else "metric")
    }
    dao.upsertBoardSetting(BoardSettingEntity(boardId, "alertPreset", selection.toString(), now))
    regenerate(dao, boardId, metric)
    customized.forEach { dao.upsertAlertRule(it) }
  }

  private suspend fun resolveBase(dao: TelemetryDao, boardId: String, fieldId: String?): Double? {
    fun values(raw: String?) = raw?.let { jsonValue(JSONObject(it)) as? Map<String, Any> } ?: emptyMap()
    return resolveConfigRelativeBase(fieldId, values(dao.getLatestBoardConfigValues(boardId)?.valuesJson), values(dao.getLatestMotorConfigValues(boardId)?.valuesJson))
  }

  suspend fun generate(dao: TelemetryDao, boardId: String, metric: String, level: String, settings: Map<String, Any?>): List<AlertRuleEntity> {
    if (level !in setOf("safe", "normal", "minimal")) return emptyList()
    val config = definition.getJSONObject("metrics").getJSONObject(metric)
    val match = if (json(settings["matchBoardConfig"]).optBoolean(metric)) definition.getJSONObject("match").optJSONObject(metric) else null
    val chosen = match ?: config
    val field = match?.getString("fieldId")
    val base = if (field != null) resolveBase(dao, boardId, field) else null
    if (match == null && config.optBoolean("requiresBatteryConfig") && !validBattery(settings["batteryConfig"])) return emptyList()
    fun rule(index: Int, start: Double, end: Double?, repeat: Long?) = AlertRuleEntity(
      boardId = boardId, id = "preset:$metric:$index", controlId = metric,
      threshold = if (field == null) start else base?.let { roundTenth(it + start) } ?: 0.0,
      thresholdMax = if (field == null) end else base?.let { b -> end?.let { roundTenth(b + it) } },
      thresholdKind = if (field == null) "fixed" else "config-relative", configFieldId = field,
      thresholdOffset = if (field == null) null else start, thresholdMaxOffset = if (field == null) null else end,
      enabled = true, soundType = config.getString("soundType"), createdAt = System.currentTimeMillis(), repeatEverySeconds = repeat, source = "preset",
    )
    if (chosen.getString("family") == "discrete") {
      val points = chosen.getJSONObject("levels").getJSONArray(level)
      return (0 until points.length()).map { index ->
        val point = points.getJSONObject(index)
        rule(index, point.getDouble("threshold"), null, if (point.isNull("repeatEverySeconds")) null else point.getLong("repeatEverySeconds"))
      }
    }
    val range = chosen.getJSONObject("levels").getJSONObject(level)
    var start = range.getDouble("start")
    var end = range.getDouble("ceiling")
    if (match == null && config.optBoolean("scaledByTopSpeed")) {
      val top = (settings["topSpeedKmh"] as? Number)?.toDouble()?.takeIf { it.isFinite() && it > 0 } ?: 50.0
      start = roundTenth(start * top); end = roundTenth(end * top)
      if (json(settings["alertPreset"]).optString("speedUnitSystem") == "imperial") {
        end = maxOf(1.0, minOf(floor(top / 1.609344), floor(end / 1.609344 + 0.5)))
        start = maxOf(0.0, minOf(end - 1, floor(start / 1.609344 + 0.5)))
        start *= 1.609344; end *= 1.609344
      }
    }
    return listOf(rule(0, start, end, null))
  }

  private val cellPresetIds by lazy {
    val data = JSONObject(checkNotNull(javaClass.getResourceAsStream("/cell-presets.json")).bufferedReader().use { it.readText() })
    val presets = data.getJSONArray("cells")
    (0 until presets.length()).map { presets.getJSONObject(it).getString("id") }.toSet()
  }
  private fun validBattery(raw: Any?): Boolean {
    val config = json(raw)
    return when (config.optString("mode")) {
      "manual" -> config.optDouble("maxVoltage") > config.optDouble("minVoltage")
      "preset" -> config.optString("cellPresetId") in cellPresetIds
      else -> false
    }
  }
}

File: modules/vescape-core/ios/telemetry/AlertPresetPersistence.swift
import Foundation
import GRDB

/// Native-owned preset intent, persisted with its rules in one transaction.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/AlertPresetPersistence.kt
final class AlertPresetPersistence {
  let writer: any DatabaseWriter
  init(writer: any DatabaseWriter) { self.writer = writer }
  static let metrics = AlertPresetGenerator.metrics
  private static let levels = AlertPresetGenerator.levels
  enum InvalidPreset: Error { case invalidIntent, missingBoard }

  static func preview(
    metric: String,
    level: String,
    topSpeedKmh: Double?,
    hasBatteryConfig: Bool,
    speedUnitSystem: String
  ) throws -> [PersistedAlertRule] {
    let input = PresetInput(topSpeedKmh: topSpeedKmh ?? 50, hasBatteryConfig: hasBatteryConfig, speedUnitSystem: speedUnitSystem)
    let specs = try AlertPresetGenerator.generate(metric: metric, level: level, input: input)
    return records(boardId: "", metric: metric, specs: specs, createdAt: 0)
  }

  private static func object(_ value: Any?) -> [String: Any] { value as? [String: Any] ?? [:] }
  private static func decode(_ json: String) throws -> Any {
    try JSONSerialization.jsonObject(with: Data(json.utf8), options: .fragmentsAllowed)
  }
  static func settings(_ db: Database, _ boardId: String) throws -> [String: Any] {
    try Dictionary(uniqueKeysWithValues: PersistedBoardSetting.filter(Column("board_id") == boardId).fetchAll(db).map { ($0.key, try decode($0.valueJson)) })
  }
  private static func save(_ db: Database, _ boardId: String, _ key: String, _ value: Any) throws {
    let data = try JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed, .sortedKeys])
    try PersistedBoardSetting(boardId: boardId, key: key, valueJson: String(decoding: data, as: UTF8.self), updatedAt: Int64(Date().timeIntervalSince1970 * 1000)).save(db)
  }
  /// Legacy JS omitted unresolved relations; repair missing slots without replacing existing rules.
  func repairMissingRelations(boardId: String) throws -> Bool {
    try writer.write { try Self.repairMissingRelations($0, boardId: boardId) }
  }
  private static func repairMissingRelations(_ db: Database, boardId: String) throws -> Bool {
    let values = try settings(db, boardId)
    let selection = object(values["alertPreset"])
    let match = object(values["matchBoardConfig"])
    let ids = Set(try PersistedAlertRule.filter(Column("board_id") == boardId).fetchAll(db).map(\.id))
    var changed = false
    for metric in metrics where match[metric] as? Bool == true {
      for rule in try generate(db, boardId: boardId, metric: metric, level: selection[metric] as? String ?? "off", settings: values) {
        if rule.thresholdKind == "config-relative", !ids.contains(rule.id) {
          try rule.save(db)
          changed = true
        }
      }
    }
    return changed
  }
  static func regenerateChanged(_ db: Database, boardId: String, before: [String: Any]) throws {
    let after = try settings(db, boardId)
    for metric in metrics where inputsChanged(metric, before: before, after: after) {
      try regenerate(db, boardId: boardId, metric: metric)
    }
  }

  private static func inputsChanged(_ metric: String, before: [String: Any], after: [String: Any]) -> Bool {
    let previous = object(before["alertPreset"])
    let next = object(after["alertPreset"])
    if (previous[metric] as? String ?? "off") != (next[metric] as? String ?? "off") { return true }
    let oldMatch = object(before["matchBoardConfig"])[metric] as? Bool ?? false
    let newMatch = object(after["matchBoardConfig"])[metric] as? Bool ?? false
    if oldMatch != newMatch { return true }
    switch metric {
    case "speed":
      let oldTop = (before["topSpeedKmh"] as? NSNumber)?.doubleValue
      let newTop = (after["topSpeedKmh"] as? NSNumber)?.doubleValue
      return oldTop != newTop
    case "battery":
      return !NSDictionary(dictionary: object(before["batteryConfig"])).isEqual(to: object(after["batteryConfig"]))
    default: return false
    }
  }

  static func regenerate(_ db: Database, boardId: String, metric: String? = nil) throws {
    let values = try settings(db, boardId)
    let selection = object(values["alertPreset"])
    for metric in metric.map({ [$0] }) ?? metrics {
      try db.execute(sql: "DELETE FROM alerts WHERE board_id = ? AND control_id = ? AND source = 'preset'", arguments: [boardId, metric])
      for rule in try generate(db, boardId: boardId, metric: metric, level: selection[metric] as? String ?? "off", settings: values) { try rule.save(db) }
    }
  }
  func apply(
    boardId: String,
    metric: String,
    action: String,
    level: String?,
    matchBoardConfig: Bool?
  ) throws {
    try writer.write { db in
      guard Self.metrics.contains(metric) else { throw InvalidPreset.invalidIntent }
      guard let board = try PersistedBoard.fetchOne(db, key: boardId), board.deletedAt == nil else { throw InvalidPreset.missingBoard }
      let values = try Self.settings(db, boardId)
      var selection = Self.object(values["alertPreset"])
      for key in Self.metrics where !Self.levels.contains(selection[key] as? String ?? "") { selection[key] = "off" }
      var customized: [PersistedAlertRule] = []
      switch action {
      case "select":
        guard let level, Self.levels.contains(level), level != "custom" else { throw InvalidPreset.invalidIntent }
        selection[metric] = level
      case "customize":
        _ = try Self.repairMissingRelations(db, boardId: boardId)
        if selection[metric] as? String == "custom" { return }
        customized = try Self.materializeRules(db, boardId: boardId, metric: metric)
        selection[metric] = "custom"
      case "discard-custom":
        try db.execute(sql: "DELETE FROM alerts WHERE board_id = ? AND control_id = ? AND (source IS NULL OR source != 'preset')", arguments: [boardId, metric])
        selection[metric] = "normal"
      case "match-board-config":
        guard let matchBoardConfig, try AlertPresetGenerator.matchField(metric) != nil else { throw InvalidPreset.invalidIntent }
        var match = Self.object(values["matchBoardConfig"])
        match[metric] = matchBoardConfig
        try Self.save(db, boardId, "matchBoardConfig", match)
      default: throw InvalidPreset.invalidIntent
      }
      try Self.save(db, boardId, "alertPreset", selection)
      try Self.regenerate(db, boardId: boardId, metric: metric)
      for rule in customized { try rule.save(db) }
    }
  }
  private static func materializeRules(_ db: Database, boardId: String, metric: String) throws -> [PersistedAlertRule] {
    let saved = try PersistedAlertRule
      .filter(Column("board_id") == boardId && Column("control_id") == metric && Column("source") == "preset")
      .fetchAll(db)
    return try saved.compactMap { rule in
      var threshold = rule.threshold
      var thresholdMax = rule.thresholdMax
      if rule.thresholdKind == "config-relative" {
        guard let base = try resolveBase(db, boardId, rule.configFieldId), let offset = rule.thresholdOffset else { return nil }
        threshold = base + offset
        thresholdMax = rule.thresholdMaxOffset.map { base + $0 }
      }
      return PersistedAlertRule(
        boardId: boardId, id: UUID().uuidString, controlId: metric,
        threshold: threshold, thresholdMax: thresholdMax, enabled: rule.enabled,
        soundType: rule.soundType, createdAt: rule.createdAt,
        repeatEverySeconds: rule.repeatEverySeconds, beepCount: rule.beepCount,
        source: nil, thresholdKind: "fixed", configFieldId: nil,
        thresholdOffset: nil, thresholdMaxOffset: nil
      )
    }
  }

  private static func resolveBase(_ db: Database, _ boardId: String, _ field: String?) throws -> Double? {
    func values(_ table: String) throws -> [String: Any] {
      let json = try String.fetchOne(db, sql: "SELECT values_json FROM \(table) WHERE board_id = ? ORDER BY captured_at DESC LIMIT 1", arguments: [boardId])
      return try json.map { object(try decode($0)) } ?? [:]
    }
    return try resolveConfigRelativeBase(field, refloat: values("board_config_values"), motor: values("motor_config_values"))
  }
  static func generate(
    _ db: Database,
    boardId: String,
    metric: String,
    level: String,
    settings: [String: Any]
  ) throws -> [PersistedAlertRule] {
    let rawUnits = try PersistedAppSetting.fetchOne(db, key: "unitSystem")
    let units = try rawUnits.map { try decode($0.valueJson) } as? String
    let matched = object(settings["matchBoardConfig"])[metric] as? Bool == true
    let field = matched ? try AlertPresetGenerator.matchField(metric) : nil
    let input = PresetInput(
      topSpeedKmh: (settings["topSpeedKmh"] as? NSNumber)?.doubleValue ?? 50,
      hasBatteryConfig: try AlertPresetGenerator.validBattery(settings["batteryConfig"]),
      speedUnitSystem: validUnitSystem(units) ?? "metric",
      matchBoardConfig: matched,
      configBase: try field.flatMap { try resolveBase(db, boardId, $0) }
    )
    let savedLevel = levels.contains(level) ? level : "off"
    let specs = try AlertPresetGenerator.generate(metric: metric, level: savedLevel, input: input)
    return records(boardId: boardId, metric: metric, specs: specs, createdAt: Int64(Date().timeIntervalSince1970 * 1000))
  }

  private static func records(
    boardId: String,
    metric: String,
    specs: [GeneratedPresetRule],
    createdAt: Int64
  ) -> [PersistedAlertRule] {
    specs.enumerated().map { index, spec in
      PersistedAlertRule(
        boardId: boardId,
        id: "preset:\(metric):\(index)",
        controlId: metric,
        threshold: spec.point.threshold,
        thresholdMax: spec.point.thresholdMax,
        enabled: true,
        soundType: spec.soundType,
        createdAt: createdAt,
        repeatEverySeconds: spec.point.repeatEverySeconds,
        beepCount: 3,
        source: "preset",
        thresholdKind: spec.relation == nil ? "fixed" : "config-relative",
        configFieldId: spec.relation?.fieldId,
        thresholdOffset: spec.relation?.offset,
        thresholdMaxOffset: spec.relation?.maxOffset
      )
    }
  }
}
